corehq/apps/userreports/reports/builder/forms.py

```diff
@@ -0,0 +1,569 @@
+            return make_form_data_source_filter(self.source_xform.data_node.tag_xmlns)
+
+    @property
+    @memoized
+    def indicators(self):
+        """
+        Return all the dict data source indicator configurations that could be
+        used by a report that uses the same case type/form as this DataSourceConfiguration.
+        """
+        ret = []
+        for prop in self.data_source_properties.values():
+            if prop['type'] == 'meta':
+                ret.append(make_form_meta_block_indicator(
+                    prop['source'][0], prop['source'][1]
+                ))
+            elif prop['type'] == "question":
+                ret.append(make_form_question_indicator(
+                    prop['source'], prop['column_id']
+                ))
+            elif prop['type'] == 'case_property':
+                ret.append(make_case_property_indicator(
+                    prop['source'], prop['column_id']
+                ))
+        ret.append({
+            "display_name": "Count",
+            "type": "count",
+            "column_id": "count"
+        })
+        return ret
+
+    @property
+    @memoized
+    def data_source_properties(self):
+        """
+        A dictionary containing the various properties that may be used as indicators
+        or columns in the data source or report.
+
+        Keys are strings that uniquely identify properties.
+        Values are dicts representing the properties, ex:
+
+        >> self.data_source_properties
+        {
+            "/data/question1": {
+                "type": "question",
+                "id": "/data/question1",
+                "text": "Enter the child's name",
+                "column_id": "data--question1",
+                "source": {
+                    'repeat': None,
+                    'group': None,
+                    'value': '/data/question1',
+                    'label': 'question1',
+                    'tag': 'input',
+                    'type': 'Text'
+                }
+            },
+            "meta/deviceID": {
+                "type": "meta",
+                "id": "meta/deviceID",
+                "text": "deviceID",
+                "column_id": "meta--deviceID",
+                "source": ("deviceID", "string")
+            }
+        }
+
+        "id" is used as the value in selects/select2s in the form. Uniquely identifies questions.
+        "column_id" is used as the column name for this indicator. There are bugs
+        with slashes which requires this to be different from "id"
+        "text" will be used as the visible text in selects/select2s
+        "type" is "question", "case_property", or "meta"
+        For questions, "source" is the dict returned by Xform.get_questions, for
+        case properties and form metadata it is simply the name of the property.
+        """
+
+        def escape_id(id):
+            # TODO: This is fairly naive escaping
+            return id.strip("/").replace("/", "__")
```

Comment:
can use something like this: https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/userreports/sql.py#L80